### ISSUE TITLE
[SURE-5877] valuesFrom does not work, neither secretKeyRef nor configMapKeyRef

### ISSUE THREAD

**Issue by papanito** [rancher/fleet#1313]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

I am using secretKeyRef but seems it's not picked up by fleet.

```
helm:
  valuesFrom:
    - secretKeyRef:
        name: clusterconfig
        namespace: cattle-gatekeeper-system
        key: values.yaml
```
Whereas the secret `clusterconfig` is in ns `cattle-gatekeeper-system` in the target (downstream) cluster.

My `fleet.yaml` looks as follows:

```
defaultNamespace: cattle-gatekeeper-system
namespace: cattle-gatekeeper-system

helm:
  chart: gatekeeper-policy-manager
  releaseName: gatekeeper-policy-manager
  version: v0.4.0

  # `config` section which contains sensitive data
  valuesFrom:
  - secretKeyRef:
      name: clusterconfig
      namespace: cattle-gatekeeper-system
      key: values.yaml

  values:
    replicaCount: 1
....
```

Secret `clusterconfig`

```
apiVersion: v1
data:
  values.yaml: XXXXXXXXXXX
kind: Secret
metadata:
  name: clusterconfig
  namespace: cattle-gatekeeper-system
type: Opaque
```

I tried it on these fleet versions

- 100.2.0+up0.5.1
- 100.0.5+up0.3.11

### Expected Behavior

Values in the secret are read and used for the helm installation

### Steps To Reproduce

1. Manually a yaml-file with `config` which contains secret information

    ```yaml
    config:
      preferredURLScheme: https
      logLevel: info
      secretKey: xxxxxx
      multiCluster:
        enabled: true
        kubeconfig: |
        apiVersion: v1
        kind: Config
        clusters:
        - name: "test"
          cluster:
          server: "https://rancher.intra/k8s/clusters/c-xxxx"
          certificate-authority-data: xxxxx
        - name: "dev"
          cluster:
          server: "https://rancher.intra/k8s/clusters/c-xxxx"
          certificate-authority-data: "xxxx"

        users:
        - name: "test"
          user:
          token: "xxxx"
        - name: "dev"
          user:
          token: "xxxx"
       
        contexts:
        - name: "test"
          context:
          user: "test"
          cluster: "test"
        - name: "dev"
          context:
          user: "dev"
          cluster: "dev"

        current-context: "dev"
    ```

2. Add it as `clusterconfig` to ns `cattle-gatekeeper-system`

    ```shell
    context=nop
    kubectl create secret generic clusterconfig \
        --from-file=values.yaml=values.$context.secret -n cattle-gatekeeper-system --context $context
    ```

3. Create a `fleet.yaml` looks as follows:

```
defaultNamespace: cattle-gatekeeper-system
namespace: cattle-gatekeeper-system

helm:
  chart: gatekeeper-policy-manager
  releaseName: gatekeeper-policy-manager
  version: v0.4.0

  # `config` section which contains sensitive data
  valuesFrom:
  - secretKeyRef:
      name: clusterconfig
      namespace: cattle-gatekeeper-system
      key: values.yaml

  values:
    replicaCount: 1
....
```

4. Add the `fleet.yaml` to the GitRepo config so that it is deployed
5. The installation fails cause the deployment complains about a missing value `secretKey` which should be read from the referenced secred

### Environment

```markdown
- Architecture: amd64
- Fleet Version: 100.2.0+up0.5.1
- Cluster:
  - Provider: RKE
  - Kubernetes Version: 1.23
```

### Anything else?

- Seems related to https://github.com/rancher/fleet/issues/724
- Manual installation of the same helm chart works fine
- Installation works fine if I add the values directly to `fleet.yaml`

**Comment by manno** [comment]
> Whereas the secret clusterconfig is in ns cattle-gatekeeper-system in the target (downstream) cluster.

The secrets need to exist in the upstream cluster. We recently improved our docs to mention this: https://fleet.rancher.io/gitrepo-structure#using-valuesfrom

If it's important to have the secrets on the downstream cluster it might be valuable to know, that we also removed the incomplete "SOPS" support from fleet. Now it should be possible to use an external SOPS controller to modify the deployments.

**Comment by manno** [comment]
I'm following the documentation: https://fleet.rancher.io/gitrepo-structure#using-valuesfrom

I came up with this example https://github.com/manno/fleet-experiments/blob/valuesFrom/valuechart/fleet.yaml and it seems to work.
I created the secret with `kubectl create secret -n default generic mysecret --from-file=values.yaml=secretdata.yaml`.

> Watch the error ;-)

What is the error :) ? You should see something in the status of the bundledeployment and in the agent's log.

**Comment by papanito** [comment]
> What is the error :) ? You should see something in the status of the bundledeployment and in the agent's log

I meant the error of the deployment which in my face complains of a mandatory value which is missing, defined in the referenced secret.

So far I did not see any word on the logs, I will check check again on Monday. I also opened a super case and showed it to one of your support guys.

**Comment by manno** [comment]
Fleet has a [mergeValues func](https://github.com/rancher/fleet/blob/master/pkg/helmdeployer/deployer.go#L753-L780) that will prefer `valuesFrom` over `values`. 

Should translate to this:
* if the key from `valuesFrom` is new, it will be set	
* it's not new:
* if the key is not a map, it will overwrite an existing value
* it's a map:
* if the existing key is not a map, overwrite it
* if the existing key is a map, too:
* recursively merge them
Maybe the key is lost while merging the values?

@papanito is there anything more which you can tell me about your use case? Do you have another `config:` key in `values:`? Do you use target customizations?

Any error would be in the agent's log on the downstream cluster. Or the bundledeployment resource's status on the upstream cluster.

**Comment by papanito** [comment]
@manno can you access Support Case 00523274? The logs are there, can't share it here

**Comment by papanito** [comment]
> @papanito is there anything more which you can tell me about your use case? 

gatekeeper-policy-manager is new, so we did not install it previously.

> Do you use target customizations?

Yes I do, I 

```
targetCustomizations:
- name: cluster1
  helm:
    values:
      ingress:
        hosts:
          - host: gpm1.intra
        paths:
        - /
        tls:
          - secretName: cert
            hosts:
              -  gpm1.intra
  clusterSelector:
    matchLabels:
      management.cattle.io/cluster-name: c-xxxxxx
- name: cluster2
  helm:
    values:
      ingress:
        hosts:
          - host: gpm2.intra
        paths:
        - /
        tls:
          - secretName: cert
            hosts:
              -  gpm1.intra
  clusterSelector:
    matchLabels:
      management.cattle.io/cluster-name: c-yyyyy
- name: cluster3
  helm:
    values:
      ingress:
        hosts:
          - host: gpm3.intra
        paths:
        - /
        tls:
          - secretName: cert
            hosts:
              -  gpm3.intra
  clusterSelector:
    matchLabels:
      management.cattle.io/cluster-name: c-zzzzz
```

**Comment by kkaempf** [comment]
> @manno can you access Support Case 00523274? The logs are there, can't share it here

That's inaccessible for engineering. Please share via JIRA.

**Comment by manno** [comment]
Thanks, I can see in the tooling YAML that `config.secretKey` is in both, `values` and the YAML from `valuesFrom`.
In my tests, `valuesFrom` would override values from `config`.

**Comment by manno** [comment]
I'm not able to reproduce this. I found however one problem  with the documentation. There is currently a bug when using both, config map and secret references for `valuesFrom` in the same list entry:

This example would ignore the configmap ref:
```
helm:
  chart: simple-chart
  valuesFrom:
    - secretKeyRef:
        name: secret-values
        namespace: default
        key: values.yaml
      configMapKeyRef:
        name: configmap-values
        namespace: default
        key: values.yaml
```

This example will use both references.
```
helm:
  chart: simple-chart
  valuesFrom:
    - secretKeyRef:
        name: secret-values
        namespace: default
        key: values.yaml
    - configMapKeyRef:
        name: configmap-values
        namespace: default
        key: values.yaml
```

**Comment by manno** [comment]
I tried again, with the opa-gatekeeper example and it deploys for me, with the correct secret key. 

Example: https://github.com/manno/fleet-experiments/blob/valuesFrom/opa-gatekeeper.yaml

**Comment by manno** [comment]
The issue here is caused by https://github.com/rancher/fleet/issues/1101

To calculate a bundle summary for the status, the fleet-controller runs a "helm template" on the chart. However, that happens in the upstream cluster and has several limitations. For example, it can't use `valuesFrom`, because the secret is in the downstream cluster.

This bug here occurs even when the secret is in the upstream cluster, though. Because the [helm template code](https://github.com/rancher/fleet/blob/master/pkg/helmdeployer/deployer.go#L437-L441)  does not load secrets at all. 

With the removal of "helm template" errors from the upstream fleet-controller, this bug should be fixed in 2.7.2 already with Fleet 0.6.0.

**Comment by sowmyav27** [comment]
@kkaempf Could we get this info for QA to start looking at this issue?
- PR fixing this issue
- QA Test template - including tests automated

**Comment by kkaempf** [comment]
Citing https://github.com/rancher/fleet/issues/1313#issuecomment-1490401256
> The issue here is fixed by https://github.com/rancher/fleet/issues/1101
PR linked in #1101: #1103

Since #1101 was already fixed in 2.7.2, I believe we can close this issue.

**Comment by sowmyav27** [comment]
Closing this issue as per this comment - https://github.com/rancher/fleet/issues/1313#issuecomment-1523120261
